src/c_types.cr

```diff
@@ -0,0 +1,151 @@
+#
+#   fun foo(x : Int) : Int     # okay
+#   fun bar(x : CTypes::Char*) # okay
+# end
+# ```
+module CTypes
+  {% unless flag?(:bits32) || flag?(:bits64) %}
+    {% if target = Crystal.constant("TARGET_TRIPLE") %}
+      {% raise "Unsupported target: #{target.id}" %}
+    {% else %}
+      # this is for old compilers that attempt to directly use a newer stdlib
+      # for cross-compilation
+      {% raise "Architecture with unsupported word size" %}
+    {% end %}
+  {% end %}
+
+  # The C `char` type.
+  #
+  # Guaranteed to have at least 8 bits. Equivalent to `UInt8` on all currently
+  # supported targets.
+  alias Char = UInt8
+
+  # The C `signed char` type.
+  #
+  # Guaranteed to have at least 8 bits. Equivalent to `Int8` on all currently
+  # supported targets.
+  alias SChar = Int8
+
+  # The C `unsigned char` type.
+  #
+  # Guaranteed to have at least 8 bits. Equivalent to `UInt8` on all currently
+  # supported targets.
+  alias UChar = UInt8
+
+  # The C `short` type.
+  #
+  # Guaranteed to have at least 16 bits. Equivalent to `Int16` on all currently
+  # supported targets.
+  alias Short = Int16
+
+  # The C `unsigned short` type.
+  #
+  # Guaranteed to have at least 16 bits. Equivalent to `UInt16` on all currently
+  # supported targets.
+  alias UShort = UInt16
+
+  # The C `int` type.
+  #
+  # Guaranteed to have at least 16 bits. Equivalent to `Int32` on all currently
+  # supported targets.
+  alias Int = Int32
+
+  # The C `unsigned int` type.
+  #
+  # Guaranteed to have at least 16 bits. Equivalent to `UInt32` on all currently
+  # supported targets.
+  alias UInt = UInt32
+
+  # The C `long` type.
+  #
+  # Guaranteed to have at least 32 bits. Equivalent to `Int32` on Windows and
+  # on 32-bit targets. Equivalent to `Int64` on 64-bit, non-Windows targets.
+  {% if flag?(:bits32) || flag?(:win32) %}
+    alias Long = Int32
+  {% elsif flag?(:bits64) %}
+    alias Long = Int64
+  {% end %}
+
+  # The C `unsigned long` type.
+  #
+  # Guaranteed to have at least 32 bits. Equivalent to `UInt32` on Windows and
+  # on 32-bit targets. Equivalent to `UInt64` on 64-bit, non-Windows targets.
+  {% if flag?(:bits32) || flag?(:win32) %}
+    alias ULong = UInt32
+  {% elsif flag?(:bits64) %}
+    alias ULong = UInt64
+  {% end %}
+
+  # The C `long long` type.
+  #
+  # Guaranteed to have at least 64 bits. Equivalent to `Int64` on all currently
+  # supported targets.
+  alias LongLong = Int64
+
+  # The C `unsigned long long` type.
+  #
+  # Guaranteed to have at least 64 bits. Equivalent to `UInt64` on all currently
+  # supported targets.
+  alias ULongLong = UInt64
+
+  # The C `float` type.
+  #
+  # Equivalent to `Float32` on all currently supported targets.
+  alias Float = Float32
+
+  # The C `double` type.
+  #
+  # Equivalent to `Float64` on all currently supported targets.
+  alias Double = Float64
+
+  # The C `intptr_t` type.
+  #
+  # Large enough to hold the value of `Pointer#address`. Equivalent to `Int32`
```

Comment:
**question:** Is this actually true? Isn't `Pointer#address` currently 64-bits on all architectures?

Reply:
It is more like "`CTypes::SizeT.new(ptr.address)` should never raise `OverflowError`" on any target. And this should be true even if `ptr = Pointer(Void).new(UInt64::MAX)` on a 32-bit platform, as `LibLLVM.build_int2ptr` will zero-extend or truncate the address